File: .claude/hooks/lib/changed-files.mjs
// @hookstack lib-changed-files
// Helper partagé par les hooks de fin de session (run-tests, stop-pytest,
// stop-quality-check, stop-duplication-check…).
//
// Le bug qu'il corrige : un hook qui décide de tourner en lisant uniquement
// `git status --porcelain` se désactive silencieusement quand l'arbre de travail
// redevient propre — par exemple après un commit (ou un push) en cours de session.
// Or les fichiers modifiés « en attente de validation » incluent aussi les
// commits déjà faits sur la branche depuis le merge-base avec origin/main.
// Ce helper combine les deux sources (union dédupliquée), comme le faisait déjà
// missing-test-detection.mjs — et retourne null hors dépôt git pour que les
// hooks gardent leur comportement historique (analyser quand même).
import { execSync } from "node:child_process";

/** Exécute une commande git et retourne sa sortie ("" si échec — jamais de throw). */
function defaultExec(cmd, opts = {}) {
	try {
		return execSync(cmd, {
			encoding: "utf8",
			cwd: opts.cwd,
			timeout: 10_000,
			stdio: "pipe",
			shell: true,
		}).trim();
	} catch {
		return "";
	}
}

/** Exécution tolérante : "" si la commande échoue (ou retourne null). */
function safe(exec, cmd, cwd) {
	try {
		const out = exec(cmd, { cwd });
		return (out ?? "").trim();
	} catch {
		return "";
	}
}

/** Parcourt la sortie de `git status --porcelain` (renames → cible). */
export function parsePorcelain(out) {
	return (out ?? "")
		.split("\n")
		.filter(Boolean)
		.map((l) => {
			const p = l.slice(3);
			return p.includes(" -> ") ? p.split(" -> ").pop() : p;
		});
}

/**
 * Fichiers modifiés à considérer pour la session :
 *  1. l'arbre de travail (staged + unstaged + untracked) via `git status --porcelain`
 *  2. les commits déjà faits sur la branche depuis le merge-base avec origin/main
 *     (le cas « worktree propre mais travail non livré » qui échappait aux hooks)
 *
 * Retourne un tableau dédupliqué trié, ou null hors dépôt git (les hooks
 * conservent alors leur comportement historique : tout analyser).
 */
export function changedFiles({
	exec = defaultExec,
	cwd = process.cwd(),
} = {}) {
	let porcelain;
	try {
		porcelain = exec("git status --porcelain", { cwd });
	} catch {
		return null; // hors dépôt git (ou git absent) → comportement historique
	}
	if (porcelain === null || porcelain === undefined) return null;

	const paths = parsePorcelain(porcelain);

	// Commits locaux déjà faits : diff contre le merge-base avec origin/main.
	const base = safe(exec, "git merge-base origin/main HEAD", cwd);
	const head = safe(exec, "git rev-parse HEAD", cwd);
	if (base && base !== head) {
		const committed = safe(exec, `git diff --name-only ${base} HEAD`, cwd);
		for (const f of (committed ?? "").split("\n").filter(Boolean)) paths.push(f);
	}

	return [...new Set(paths)].sort();
}

File: .claude/hooks/lib/changed_files.py
# @hookstack lib-changed-files
"""Shared helper for end-of-session hooks (stop-pytest, stop-quality-check,
stop-duplication-check...). Python twin of lib/changed-files.mjs.

The bug it fixes: a hook deciding to run by reading only ``git status
--porcelain`` silently disables itself once the working tree is clean again —
e.g. after a commit (or push) mid-session. Files "pending validation" also
include commits already made on the branch since the merge-base with
origin/main. This helper combines both sources (deduplicated union), like
missing-test-detection already did — and returns None outside a git repo so
hooks keep their historical behaviour (analyse anyway).
"""
import subprocess

TIMEOUT = 10


def _run(cmd, *, cwd=None):
    """Runs a git command; returns its trimmed stdout, or "" on failure (never raises)."""
    try:
        out = subprocess.run(
            cmd,
            shell=True,
            capture_output=True,
            text=True,
            timeout=TIMEOUT,
            cwd=cwd,
            check=True,
        ).stdout
        return (out or "").strip()
    except Exception:
        return ""


def _safe(exec_cmd, cmd):
    """Tolerant run: "" if the command fails (or returns a falsy value)."""
    try:
        out = exec_cmd(cmd)
        return (out or "").strip()
    except Exception:
        return ""


def _parse_porcelain(out):
    """Parses ``git status --porcelain`` output (renames → target)."""
    paths = []
    for line in (out or "").splitlines():
        p = line[3:]
        paths.append(p.split(" -> ")[-1] if " -> " in p else p)
    return paths


def changed_files(*, exec_cmd=None, cwd=None):
    """Changed files for the session:
      1. the working tree (staged + unstaged + untracked) via porcelain
      2. commits already made on the branch since the merge-base with origin/main
         (the "clean tree but unshipped work" case that used to escape the hooks)

    Returns a deduplicated sorted list, or None outside a git repo (hooks keep
    their historical behaviour: analyse everything).
    """
    if exec_cmd is None:
        exec_cmd = lambda cmd: _run(cmd, cwd=cwd)
    try:
        porcelain = exec_cmd("git status --porcelain")
    except Exception:
        return None  # outside a git repo (or git missing) → historical behaviour
    if porcelain is None:
        return None

    paths = _parse_porcelain(porcelain)

    # Local commits already made: diff against the merge-base with origin/main.
    base = _safe(exec_cmd, "git merge-base origin/main HEAD")
    head = _safe(exec_cmd, "git rev-parse HEAD")
    if base and base != head:
        committed = _safe(exec_cmd, f"git diff --name-only {base} HEAD")
        paths += [f for f in (committed or "").splitlines() if f.strip()]

    return sorted(set(paths))
